images/bridge2/translate/translate.go

```diff
@@ -0,0 +1,112 @@
+package translate
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/ajbouh/substrate/images/bridge2/tracks"
+	"github.com/ajbouh/substrate/images/bridge2/transcribe"
+)
+
+var recordTranslation = tracks.EventRecorder[*TranslationEvent]("translation")
+
+type TranslationEvent struct {
+	SourceEvent tracks.ID
+	Translation *Translation
+}
+
+type Agent struct {
+	Endpoint       string
+	TargetLanguage string
+}
+
+func (a *Agent) HandleEvent(annot tracks.Event) {
+	if annot.Type != "transcription" {
+		return
+	}
+
+	in := annot.Data.(*transcribe.Transcription)
+	// This doesn't account for fuzzy matching, so SourceLanguage
+	// comes back as "en", but the canonical for target seems to be "eng".
```

Comment:
When you say "doesn't account for", what do you mean? Sadly there doesn't seem to be a clear answer on what truly "canonical" names for languages are.

Reply:
In order for this code to detect if the text is already in the target language, I'm only relying on an exact match, but the transcription and translation services seem to have a different "canonical" representation of the language codes.
So, for this to work I have to set `TargetLanguage = "en"` since that's what the transcription service reports. However, that leads to warnings in the translator like:
```
warning: no exact language match for target language en, using eng; other possibilities were ['eng']
```
Though it may be better to handle this in the translator in some way. If we pass in something like `source = en, target = eng` it could respond to indicate that it considers those to be the same. Or if the translator response contained the normalized codes in the output, we could compare them after translation to detect if this was passed through without translation.
I guess with the blackboard model it would be up to the translator to look at each new transcription anyways, and then decide which languages to produce translations to.
So I'm not sure if we really need to do something specific with this now, but calling it out since changing the `TargetLanguage` to something else might end up producing "identity" translations when the only difference is the two variations of the language code.

I'd have to take a closer look to be sure, but I would assume they're both using ISO 639 codes, though the 2- and 3-character codes appear to be different levels of granularity:
https://en.wikipedia.org/wiki/List_of_ISO_639_language_codes

Reply:
Unfortunately the list used by seamlessm4t isn't that set of ISO language codes. I went through all this and ended up just writing the fuzzy matcher.

I solved this problem in the original implementation of bridge by having a set of user-provided target language "aliases". If a target language for a transcript matches *any* of the provided aliases then I don't bother translating. If none match, then I use the "primary" (aka first) given alias for the target language. This worked well and was very easy to implement.